[Triton] [Inductor] Enable padded stride support

njriasan opened the pull request:
Summary:
Issue I noticed while fixing tests for TMA store. This `triton.language.make_tensor_descriptor` call hardcodes the shape information as the stride, which is not necessarily correct.

In particular, its legal to have a stride bigger than the shape (e.g. padded to a size). A good example of the usage of this would be to allocate a tensor to always be a mulitple of 16 and just pad the result so TMA is legal.

Test Plan:
Tested with `buck2 run mode/opt-split-dwarf mode/inplace -c fbcode.nvcc_arch=h100 caffe2/test/inductor:max_autotune 2>&1 | tee ~/test_logs.log` and confirmed all max autotune tests passed.

Rollback Plan:

Reviewed By: NikhilAPatel

Differential Revision: D80123894

Diff train notes from clee2000:
PR didn't get merged, diff got merged and then reverted, adding tags to make ship it ignore this
@diff-train-skip-merge


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

eellison reviewed (commented):
`_is_tma_compatible` allows tensors with dims 3 and 4 as well. How are those handled ?

Also, when you're adding to `test_max_autotune`, make sure to trigger it with ciflow/h100 (it should be automatic but it's not yet).

njriasan committed:
[Triton] [Inductor] Enable padded stride support (#160493)

Summary:

Issue I noticed while fixing tests for TMA store. This `triton.language.make_tensor_descriptor` call hardcodes the shape information as the stride, which is not necessarily correct.

In particular, its legal to have a stride bigger than the shape (e.g. padded to a size). A good example of the usage of this would be to allocate a tensor to always be a mulitple of 16 and just pad the result so TMA is legal.

Test Plan:
Tested with `buck2 run mode/opt-split-dwarf mode/inplace -c fbcode.nvcc_arch=h100 caffe2/test/inductor:max_autotune 2>&1 | tee ~/test_logs.log` and confirmed all max autotune tests passed.

Rollback Plan:

Reviewed By: NikhilAPatel

Differential Revision: D80123894

njriasan commented on test/inductor/test_max_autotune.py:
```diff
@@ -165,6 +165,62 @@ def mm(a, b):
+        b = torch.empty_strided(b_shape, b_stride, dtype=torch.float16)
+        b[:] = torch.randn(b_shape, dtype=torch.float16)
+        b = b.to(GPU_TYPE)
+        with config.patch(
+            {
+                "max_autotune": True,
+                "triton.enable_persistent_tma_matmul": "1",
+                "test_configs.autotune_choice_name_regex": "mm_persistent_tma",
+            }
+        ):
+            c_actual = torch.compile(mm, dynamic=dynamic)(a, b)
+            c_expected = mm(a, b)
+
+        torch.testing.assert_close(c_actual, c_expected, atol=1e-2, rtol=1e-2)
```
Yes I can add that check.

njriasan commented:
> > `_is_tma_compatible` allows tensors with dims 3 and 4 as well. How are those handled ?
> > Also, when you're adding to `test_max_autotune`, make sure to trigger it with ciflow/h100 (it should be automatic but it's not yet).
> 
> I don't believe dims 3 and 4 are handled yet. I'll update the checks in the templates to restrict this further.

Looks like that check is missing from `use_triton_tma_template`. I'll add it.
